package eu.point.tmsdn.impl;

option java_package = "eu.point.tmsdn.impl";
option java_outer_classname = "TmSdnMessages";

message TmSdnMessage {
	enum TmSdnMessageType {
		LS = 0;
		TM = 1;
		RR = 2;
		RO = 3;
	}

	message LinkStatusMessage {
		enum LSMType {
			ADD = 0;
			RMV = 1;
		}

		required LSMType lsmType = 1;
		required string nodeID1 = 2;
		required string nodeID2 = 3;	
	}

	message TrafficMonitoringMessage {
		required string nodeID1 = 1;
		required string nodeID2 = 2;	
		required int64 packetsReceived = 3;
		required int64 packetsTransmitted = 4;	
		required int64 bytesReceived = 5;
		required int64 bytesTransmitted = 6;	
	}

	message ResourceRequestMessage {
		message RecourceRequest {
			required string srcNode = 1;
			required string dstNode = 2;
			optional string srcMac = 3;
			optional string nodeConnector = 4;
		}
		repeated RecourceRequest requests = 1;
	}
	
	message ResourceOfferMessage {
		message RecourceOffer {
			required string nid = 1;
			required string lid = 2;
			required string ilid = 3;
			optional string srcMac = 4;
		}
		repeated RecourceOffer offers = 1;
	}
	
	optional LinkStatusMessage linkStatusMessage = 1;
	optional TrafficMonitoringMessage trafficMonitoringMessage = 2;
	optional ResourceRequestMessage resourceRequestMessage = 3;
	optional ResourceOfferMessage resourceOfferMessage = 4;
	required TmSdnMessageType type = 5;
}